transformers/modeling_beam_search.py

```diff
@@ -0,0 +1,240 @@
+# coding=utf-8
+# Copyright (c) 2019 Yang Liu
+
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+A general wrapper around models with LM heads to generate sequences
+using beam search.
+"""
+import torch
+from torch import nn
+
+
+class ModelWithBeamSearch(nn.Module):
```

Comment:
Per discussion, it would be preferable to have a separate beam search class where the `step` could be repurposed for encoder-only model. The. add a `generate` function that is structure-specific and performs the full beam search.

The workflow would look like:

```python
model = Model2Model.from_pretrained(...)
search = TransformerBeamSearch(model, ...)
output = model.generate('my dog is cute')
```

Reply:
Maybe see if we could add a Decoder base class in which we can add several decoding mechanisms that are currently in `run_generation`.

Reply:
Yes, see below. The term `Decoder` is ambiguous in this context, let us find something else. `BeamSearch` and `Sampling` are explicit, and we could even put them together in a `generate` module .